sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+
+    // Path to cmd.exe
+    std::vector<CHAR> commandLineStr;
+    {
+      auto cmd = Environment::GetVariable("COMSPEC");
+      EnsureShellExists(cmd);
+
+      // Enclose path in quotes, in case there are space characters.
+      {
+        std::string const Quote = "\"";
+        cmd = Quote + cmd + Quote;
+      }
+
+      auto const commandLine = cmd + " /c " + command;
+      commandLineStr.insert(commandLineStr.end(), commandLine.begin(), commandLine.end());
+      commandLineStr.push_back('\0');
+    }
+
+    // Form the environment
+    std::vector<CHAR> environmentValues;
+    LPVOID lpEnvironment = NULL;
+    {
+      {
+        constexpr auto PathEnvVarName = "PATH";
+        auto pathValue = Environment::GetVariable(PathEnvVarName);
+
+        for (auto const pf :
+             {Environment::GetVariable("ProgramFiles"),
+              Environment::GetVariable("ProgramFiles(x86)")})
+        {
+          if (!pf.empty()) // LCOV_EXCL_LINE
+          {
+            if (!pathValue.empty()) // LCOV_EXCL_LINE
+            {
+              pathValue += ";";
+            }
+
+            pathValue += pf + "\\Microsoft SDKs\\Azure\\CLI2\\wbin";
+          }
+        }
+
+        AppendToEnvironmentValuesIfNotEmpty(environmentValues, PathEnvVarName, pathValue);
+      }
+
+      // Also provide SystemRoot variable.
+      // Without it, 'az' may fail with the following error:
+      // "Fatal Python error: _Py_HashRandomization_Init: failed to get random numbers to
+      // initialize Python
+      // Python runtime state: preinitialized
+      // ".
+      AppendToEnvironmentValuesIfDefined(environmentValues, "SystemRoot");
+
+      // Also provide USERPROFILE variable.
+      // Without it, we'll be getting "ERROR: Please run 'az login' to setup account." even if the
+      // user did log in.
+      AppendToEnvironmentValuesIfDefined(environmentValues, "USERPROFILE");
+
+      if (!environmentValues.empty()) // LCOV_EXCL_LINE
+      {
+        environmentValues.push_back('\0'); // terminate the block
+        lpEnvironment = environmentValues.data();
+      }
+    }
+
+    ThrowIfApiCallFails(
+        CreateProcessA(
+            NULL,
+            commandLineStr.data(),
+            NULL,
+            NULL,
+            TRUE,
+            NORMAL_PRIORITY_CLASS | CREATE_NO_WINDOW,
+            lpEnvironment,
+            NULL,
+            &startupInfo,
+            &procInfo),
+        "Cannot create process");
+  }
+
+  // We won't be needing the process main thread handle on our end.
+  static_cast<void>(CloseHandle(procInfo.hThread));
+
+  // Keep the process handle so we can cancel it if it takes too long.
+  m_processHandle = procInfo.hProcess;
+
+  // We won't be writing to the pipe that is meant for the process.
+  // We will only be reading the pipe.
+  // So, now that the process is started, we can close write handle on our end.
+  static_cast<void>(CloseHandle(outputPipe.m_writeHandle));
```

Comment:
Why did you need the write handle at all? You should be able to redirect output without also redirecting input.

Reply:
Write handle is not for input, it is writing end of the pipe that we use for redirecting output.
We just passed the writing end to the process that we created (lines 405 and 406), and won't be needing it on our end, so we close it.

```cpp
startupInfo.hStdOutput = outputPipe.m_writeHandle;
startupInfo.hStdError = outputPipe.m_writeHandle;
```